File: ocaml/bob/bob.ml
open Base

let is_shout str =     
    String.exists str ~f:Char.is_alpha && String.for_all str ~f:(fun c -> not (Char.is_alpha c) || Char.is_uppercase c)

let is_question phrase =
  String.is_suffix phrase ~suffix:"?"

let response_for input =
    match String.strip input with
    | s when is_question s && is_shout s -> "Calm down, I know what I'm doing!"
    | s when is_shout s -> "Whoa, chill out!"
    | s when is_question s -> "Sure."
    | s when String.is_empty s -> "Fine. Be that way!"
    | _  -> "Whatever."

File: scala/bob/src/main/scala/Bob.scala
class Bob {
  def checkUpperCase(input: String) = {
    input.map(_.toUpper) == input && input.map(_.toLower) != input
  }

  def hey(remark: String): String = remark match {
    case empty if remark forall (_.isWhitespace) => "Fine. Be that way!"
    case yell if checkUpperCase(remark) => "Whoa, chill out!"
    case question if remark.endsWith("?") => "Sure."
    case _ => "Whatever."
  }
}
